rir/src/compiler/opt/gvn.cpp

```diff
@@ -0,0 +1,204 @@
+                changed = true;
+            };
+
+            auto assignNumber = [&](Value* v) {
+                while (classes.count(nextNumber))
+                    nextNumber++;
+                storeNumber(v, nextNumber, {nextNumber});
+                return nextNumber;
+            };
+
+            auto i = Instruction::Cast(v);
+            if (!i) {
+                assignNumber(v);
+                return;
+            }
+
+            if (!i->producesRirResult())
+                return;
+
+            auto computeNumber = [&](Instruction* i) {
+                size_t klassNumber = i->gvnBase();
+
+                std::vector<size_t> args;
+                if (auto phi = Phi::Cast(i)) {
+                    bool success = false;
+                    auto res = computePhiNr(phi, success);
+                    if (success) {
+                        changed = true;
+                        number[phi] = res;
+                        if (!firstValue.count(res))
+                            firstValue[res] = phi;
+                    }
+                    return;
+                }
+
+                bool success = true;
+                i->eachArg([&](Value* a) {
+                    if (success && number.count(a))
+                        klassNumber = hash_combine(klassNumber, number.at(a));
+                    else
+                        success = false;
+                });
+                if (success) {
+                    i->eachArg([&](Value* a) { args.push_back(number.at(a)); });
+
+                    if (!classes.count(klassNumber) ||
+                        classes.at(klassNumber) != args) {
+                        while (classes.count(klassNumber))
+                            klassNumber++;
+                        storeNumber(i, klassNumber, args);
+                    } else {
+                        number[i] = klassNumber;
+                    }
+                    changed = true;
+                }
+            };
+
+            if (auto ld = LdConst::Cast(i)) {
+                SEXP constant = ld->c();
+                for (auto& c : constants) {
+                    if (R_compute_identical(c.first, constant, 0)) {
+                        number[i] = c.second;
+                        changed = true;
+                        return;
+                    }
+                }
+                while (classes.count(nextNumber))
+                    nextNumber++;
+                constants.emplace(constant, nextNumber);
+                storeNumber(i, nextNumber, {nextNumber});
+            } else {
+                // We allow instructions with those effects to be deduplicated.
+                auto effects =
+                    i->effects & (~(Effects(Effect::Error) | Effect::Warn |
+                                    Effect::Visibility | Effect::Force));
+                if (effects.empty())
+                    computeNumber(i);
+                else
+                    assignNumber(i);
```

Comment:
Why do we care at all? Why not just adding the effects check at the beggining?

Reply:
sorry, I don't understand the question. and which beginning?

Reply:
I meant...it seems that for any instruction with "real effect" we do not need to do more than `assignNumber`. You have this above in the same function:
```
auto i = Instruction::Cast(v);
            if (!i) {
                assignNumber(v);
                return;
            }
```
why not just add `|| i->effects()`

Reply:
ok, i can do that, if you prefer